Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,671 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+using MediatR;
+using Microsoft.Boogie;
+using Microsoft.Dafny.LanguageServer.Language;
+using OmniSharp.Extensions.JsonRpc;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider declaring as nullable.
+
+namespace Microsoft.Dafny.LanguageServer.Workspace.Notifications {
+  /// <summary>
+  /// DTO used to communicate the current verification diagnostics to the LSP client.
+  /// </summary>
+  [Method(DafnyRequestNames.VerificationDiagnostics, Direction.ServerToClient)]
+  public class VerificationDiagnosticsParams : IRequest, IRequest<Unit> {
+    public VerificationDiagnosticsParams(DocumentUri uri,
+      int version,
+      NodeDiagnostic[] perNodeDiagnostic,
+      Container<Diagnostic> diagnostics,
+      int linesCount,
+      bool verificationStarted,
+      int numberOfResolutionErrors) {
+      Uri = uri;
+      Version = version;
+      PerNodeDiagnostic = perNodeDiagnostic;
+      Diagnostics = diagnostics;
+      if (linesCount != 0) { // Deserialization makes linesCount to be equal to zero.
+        PerLineDiagnostic =
+          RenderPerLineDiagnostics(this, perNodeDiagnostic, linesCount, numberOfResolutionErrors, verificationStarted, diagnostics);
+      }
+    }
+
+    /// <summary>
+    /// Gets the URI of the document whose verification completed.
+    /// </summary>
+    public DocumentUri Uri { get; init; }
+
+    /// <summary>
+    /// Gets the version of the document.
+    /// </summary>
+    public int? Version { get; init; }
+
+    /// <summary>
+    /// Gets the same diagnostics as displayed in the diagnostics window
+    /// </summary>
+    public Container<Diagnostic> Diagnostics { get; init; }
+
+    /// <summary>
+    /// Returns a tree of diagnostics that can be used
+    /// First-level nodes are methods, or witness subset type checks
+    /// Second-level nodes are preconditions, postconditions, body verification status
+    /// Third level nodes are assertions inside functions 
+    /// </summary>
+    public NodeDiagnostic[] PerNodeDiagnostic { get; init; }
```

Comment:
If this is an external API, I worry that "node" is too generic a term to be widely understood here.